### ISSUE TITLE
Typecheck against a curried role dies

### ISSUE THREAD

**Issue by vrurg** [rakudo/rakudo#2698]
## The Problem

Has been discovered on a typed `Hash` definition but stems from a common issue in `accepts_type` method of `Metamodel::CurriedRoleHOW`.

When a `Hash` is parameterized with both key and value types iterating over it's roles and matching against the hash type itself fails with _"Cannot find method 'ACCEPTS' on object of type TValue"_ error. 

## Steps to Reproduce

This code demonstrates the problem:

     my \t = Hash[Any:D,Any:D];
     for t.^roles {
         use nqp;
         note $_.^name, " of ", $_.HOW.^name;
         note nqp::istype(nqp::decont(t), nqp::decont($_));
     }

`nqp::istype` is used because this is where smartmatch ends up anyway.

## Analysis 

The problem comes from the fact that upon mixin in `TypedHash` role the `Hash` also gets `Associative[TValue]` into its role list. And it gets it as is, with `TValue`, which is of `Metamodel::GenericHOW` type. `nqp::istype` refers to `accepts_type` method of `Metamodel::CurriedRoleHOW` which in turn tries `.ACCEPTS(Any:D)` on `TValue` causing the exception.

## Environment
* Compiler version (`perl6 -v`): This is Rakudo version 2018.12 built on MoarVM version 2018.12

**Comment by vrurg** [comment]
Not sure what the fix must look like, but suspect that `Associative` must be parametrized with `Any:D` (or whatever else a user supplies).

**Comment by vrurg** [comment]
Basically, the problem was tracked down to the following code:

    use nqp;
    role R0[::T] {
    }
    role R1[::T1] does R0[T1] {
    }
    my \R1-Int = R1[Int:D];
    for R1-Int.^roles {
        note " -> role: ", $_.^name, " of ", $_.HOW.^name;
        note " -> istype: ", nqp::istype(nqp::decont(R1-Int), nqp::decont($_));
    }

Which outputs:

>```
> -> role: R0[T1] of Perl6::Metamodel::CurriedRoleHOW
> -> istype: 0
>```

The problem is that `R0` is not parameterized with `Int:D` (or whatever type is passed to `R1`). With more investigation I discovered that final parameterization of `R0` takes places only when `R1` gets specialized, i.e. – when applied or mixed into a class.

I'm trying to work our a fix for the problem. In my opinion, `R0` must be specialized in order to pass `nqp::istype` check. But this better be done lazily as to avoid possible performance hit while typechecking against unspecialized role is rather rare operation. 

So far, the best place for this operation is probably `accepts_type` method of `Metamodel::CurriedRoleHOW`. But I don't know yet if there is a way to get it done properly in there.

Looking forward for any advice on this.

**Comment by vrurg** [comment]
Another user-confusing side effect happens when role gets consumed by a class:

     role R0[::T] {
         has T $.r0 is rw;
     }
     role R1[::T1] does R0[T1] {
     }
     class C does R1[Int] {
     }
     my $c = C.new;
     say "r0.^name: ", $c.r0.^name;
     say "VAR.of: ", $c.r0.VAR.of.^name;
     $c.r0 = 42;
     say "r0.^name: ", $c.r0.^name;

The output of this:

>```
>r0.^name: T
>VAR.of: Int
>r0.^name: Int
>```

**Comment by jnthn** [comment]
I think the final issue here is distinct: it's because `Attribute.instantiate_generic` doesn't do instantiation of the default value of the auto-viv container. That method is in `Perl6::Metamodel::BOOTSTRAP`. It touches the descriptor here:

```
nqp::bindattr($avc_copy, @avc_mro[$i], '$!descriptor', $cd_ins);
```

But doesn't touch `$!value`, which is why we end up with the uninstantiated `T` in there. So this particular part should be an easy fix.

**Comment by vrurg** [comment]
Unluckily to me, I wished to see something about the roles in first place. Still can't think of how it must look eventually. As a user, I would say that after parameterizing with a concrete type:

```
my \r1 = R1[Int];
```

introspection of `R1` must provide me with `R0` being parameterized with `Int` too.

My biggest confusion here is the metamodel structure. The problem arises from the fact that curried role isn't a real role and it simply proxies a real role metaobject. So that when asking for `R1[Int]` roles we actually ask underlying one of `ParametricRole*HOW`. Sure enough, it reports back roles curried with a generic argument. A typical case of lazy operation where actual instantiation/specialization is postponed until roles are applied to class. 

The problem lies in the fact that sometimes we need to deal with roles before they get specialized! I plunged into this while working on #2250 and #2677 where collecting all `submethod`s requires concretizations of class roles and in some cases a call to `nqp::istype` is dying because of failing type check against a generic role parameter.

While trying to work out a fix for this ticket I came to a conclusion that role currying must be somewhat reconsidered. `CurriedRoleHOW` object needs to have its own list of roles to apply (does `RoleContainer`) filled with particular parameterizations when the exact parameters are already known.

**Comment by vrurg** [comment]
I think I've got this one fixed the way I mentioned above: by consuming `Metamodel::RoleContainer` and implementing own `role_typecheck_list` and method `type_check`. So far, the fix is passing spectest (well, up to the point the master branch does it – i.e. both fail on same three tests for me).

Plan to submit a PR after certain clean up and implementing fix-specific roast tests.

**Comment by vrurg** [comment]
Closed via #2715

**Comment by vrurg** [comment]
Tests are provided with perl6/roast#513

**Comment by lizmat** [comment]
t/spec/S14-roles/parameterized-basic.t now has 3 test failures, is that correct ?

**Comment by vrurg** [comment]
No, it shouldn't after merging #2715. I'll check it later today.

**Comment by vrurg** [comment]
Closed via #2715 and #2932
